src/smolagents/agents.py

```diff
@@ -242,108 +242,131 @@ def _initialize_agent(
-                    for name, cls in TOOL_MAPPING.items()
-                    if name != "python_interpreter" or self.__class__.__name__ == "ToolCallingAgent"
-                }
-            )
+            for tool_name, tool_class in TOOL_MAPPING.items():
+                if tool_name != "python_interpreter" or self.__class__.__name__ == "ToolCallingAgent":
+                    self.tools[tool_name] = tool_class()
         self.tools["final_answer"] = FinalAnswerTool()
 
     def _run(self, task: str, images: List[str] | None = None) -> Generator[ActionStep | AgentType, None, None]:
-        final_answer, self.step_number = None, 1
+        """
+        Run the agent in streaming mode and returns a generator of all the steps.
+
+        Args:
+            task (`str`): Task to perform.
+            images (`list[str]`): Paths to image(s).
+        """
+        final_answer = None
+        self.step_number = 1
         while final_answer is None and self.step_number <= self.max_steps:
             step_start_time = time.time()
-            memory_step = self._create_memory_step(step_start_time, images)
+            memory_step = ActionStep(
+                step_number=self.step_number,
+                start_time=step_start_time,
+                observations_images=images,
+            )
             try:
-                final_answer = self._execute_step(task, memory_step)
+                if self.planning_interval is not None and self.step_number % self.planning_interval == 1:
+                    self.planning_step(
+                        task,
+                        is_first_step=(self.step_number == 1),
+                        step=self.step_number,
+                    )
+                self.logger.log_rule(f"Step {self.step_number}", level=LogLevel.INFO)
+
+                # Run one step!
+                final_answer = self.step(memory_step)
+                if final_answer is not None and self.final_answer_checks is not None:
+                    for check_function in self.final_answer_checks:
+                        try:
+                            assert check_function(final_answer, self.memory)
+                        except Exception as e:
+                            final_answer = None
+                            raise AgentError(f"Check {check_function.__name__} failed with error: {e}", self.logger)
             except AgentError as e:
                 memory_step.error = e
             finally:
-                self._finalize_step(memory_step, step_start_time)
-                yield memory_step
+                memory_step.end_time = time.time()
+                memory_step.duration = memory_step.end_time - step_start_time
+                self.memory.steps.append(memory_step)
+                for callback in self.step_callbacks:
+                    # For compatibility with old callbacks that don't take the agent as an argument
+                    if len(inspect.signature(callback).parameters) == 1:
```

Comment:
I think it is better to do
```py
if "agent" in inspect.signature(callback).parameters:
    callback(memory_step, agent=self)
else:
   callback(memoery_step)
```
Because there could have been callbacks with default args, for example:
```py
def my_callback(memory_step, some_optional_param=None)
   pass
```
This check will treat it as a "new style" callback and fail. All we care here is if it has `agent` kwarg, so let's check secifically for it.